File: CSharp/ExtensibleCommands/ExtensibleCommands/RetryCommand.cs
using System.Threading;

namespace ExtensibleCommands
{
    /// <summary>
    /// Retries core command N times if it fails
    /// </summary>
    public class RetryCommand : DecoratorCommand
    {
        /// <summary> Number of command retries </summary>
        public int NumberOfRetries { get; }

        /// <summary> Retry delay (in msec) </summary>
        public int RetryDelayMsec { get; }

        /// <summary> Current retry index </summary>
        public int CurrentRetryIndex { get; private set; }

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="coreCommand">Core command</param>
        /// <param name="numberOfRetries">Number of command retries</param>
        public RetryCommand(ICommand coreCommand, int numberOfRetries)
            : this(coreCommand, numberOfRetries, 0, "Retry") { }

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="coreCommand">Core command</param>
        /// <param name="numberOfRetries">Number of command retries</param>
        /// <param name="retryDelayMsec">Retry delay (in msec)</param>
        public RetryCommand(ICommand coreCommand, int numberOfRetries, int retryDelayMsec)
            : this(coreCommand, numberOfRetries, retryDelayMsec, "Retry") { }

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="coreCommand">Core command</param>
        /// <param name="numberOfRetries">Number of command retries</param>
        /// <param name="retryDelayMsec">Retry delay (in msec)</param>
        /// <param name="name">Command name</param>
        public RetryCommand(ICommand coreCommand, int numberOfRetries, int retryDelayMsec, string name)
            : base(coreCommand, name)
        {
            NumberOfRetries = numberOfRetries;
            RetryDelayMsec = retryDelayMsec;
        }

        protected override void Execute()
        {
            CurrentRetryIndex = 0;
            for (int i = 0; i < NumberOfRetries; i++)
            {
                CurrentRetryIndex++;
                CoreCommand.Run();

                ProcessAbortAndPauseEvents();

                if (CoreCommand.CurrentState == State.Completed || CoreCommand.CurrentState == State.Aborted || CurrentState == State.Aborted)
                    break;
                if (CoreCommand.CurrentState == State.Failed && !(CoreCommand.Exception is ExtensibleCommandsAllowRetryException))
                    break;

                if (CoreCommand.Exception is ExtensibleCommandsAllowRetryException)
                    Logger.Log(Logger.LogLevel.Error,
                        string.Format("ERROR (RECOVERED)[{0}] - {1}", CoreCommand.Exception.ID, CoreCommand.Exception.Text));
            }

            // If after all retries there is still error, we need to report it.
            if (CoreCommand.CurrentState == State.Failed)
            {
                CurrentState = State.Failed;
                Exception = CoreCommand.Exception;
            }

            Thread.Sleep(RetryDelayMsec);
        }
    }
}

File: Java/ExtensibleCommands/src/main/java/org/extensiblecommands/RetryCommand.java
package org.extensiblecommands;

import static java.lang.Thread.sleep;

/**
 * Retries core command N times if it fails
 */
public class RetryCommand extends DecoratorCommand {
    /**
     * Number of command retries
     */
    private final int numberOfRetries;
    /**
     * Retry delay (in msec)
     */
    private final int retryDelayMsec;
    /**
     * Current retry index
     */
    private int currentRetryIndex;

    /**
     * Constructor
     * @param coreCommand       Core command
     * @param numberOfRetries   Number of command retries
     */
    public RetryCommand(Command coreCommand, int numberOfRetries) {
        this(coreCommand, numberOfRetries, 0, "Retry");
    }

    /**
     * Constructor
     * @param coreCommand       Core command
     * @param numberOfRetries   Number of command retries
     * @param retryDelayMsec    Retry delay (in msec)
     */
    public RetryCommand(Command coreCommand, int numberOfRetries, int retryDelayMsec) {
        this(coreCommand, numberOfRetries, retryDelayMsec, "Retry");
    }

    /**
     * Constructor
     * @param coreCommand       Core command
     * @param numberOfRetries   Number of command retries
     * @param retryDelayMsec    Retry delay (in msec)
     * @param name              Command name
     */
    public RetryCommand(Command coreCommand, int numberOfRetries, int retryDelayMsec, String name) {
        super(coreCommand, name);
        this.numberOfRetries = numberOfRetries;
        this.retryDelayMsec = retryDelayMsec;
    }

    /**
     * @return          Number of command retries
     */
    public final int getNumberOfRetries() {
        return numberOfRetries;
    }

    /**
     * @return          Retry delay (in msec)
     */
    public final int getRetryDelayMsec() {
        return retryDelayMsec;
    }

    /**
     * @return          Current retry index
     */
    public final int getCurrentRetryIndex() {
        return currentRetryIndex;
    }

    @Override
    protected void execute() throws Exception {
        currentRetryIndex = 0;
        for (int i = 0; i < numberOfRetries; i++) {
            currentRetryIndex++;
            coreCommand.run();

            processAbortAndPauseEvents();

            if (coreCommand.getState() == State.Completed || coreCommand.getState() == State.Aborted || getState() == State.Aborted)
                break;
            if (coreCommand.getState() == State.Failed && !(coreCommand.getException() instanceof ExtensibleCommandsAllowRetryException))
                break;

            if (coreCommand.getException() instanceof ExtensibleCommandsAllowRetryException)
                Logger.log(Logger.LogLevel.Error,
                        String.format("ERROR (RECOVERED)[%s] - %s", getCoreCommand().getException().getId(),
                                getCoreCommand().getException().getText()));
        }

        // If after all retries there is still an error, we need to report it.
        if (coreCommand.getState() == State.Failed) {
            setState(State.Failed);
            exception = coreCommand.getException();
        }

        sleep(retryDelayMsec);
    }
}
